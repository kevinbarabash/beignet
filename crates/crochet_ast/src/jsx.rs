use crate::expr::Expr;
use crate::ident::Ident;
use crate::lit::Lit;
use crate::span::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXText {
    pub span: Span,
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXExprContainer {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXElement {
    pub span: Span,
    // Other ASTs make have JSXOpeningElement and JSXClosingElement
    pub name: String,
    pub attrs: Vec<JSXAttr>,
    pub children: Vec<JSXElementChild>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum JSXElementChild {
    JSXText(JSXText),
    JSXExprContainer(JSXExprContainer),
    JSXElement(Box<JSXElement>),
    // JSXSpreadChild(JSXSpreadChild),
    // JSXFragment(JSXFragment),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXAttr {
    pub span: Span,
    pub ident: Ident,
    pub value: JSXAttrValue,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum JSXAttrValue {
    Lit(Lit),
    JSXExprContainer(JSXExprContainer),
}
