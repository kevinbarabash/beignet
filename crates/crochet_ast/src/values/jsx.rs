use crate::values::expr::Expr;
use crate::values::ident::Ident;
use crate::values::lit::Lit;
use crate::values::span::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXText {
    pub span: Span,
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXExprContainer {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXElement {
    pub span: Span,
    // Other ASTs make have JSXOpeningElement and JSXClosingElement
    pub name: String,
    pub attrs: Vec<JSXAttr>,
    pub children: Vec<JSXElementChild>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum JSXElementChild {
    JSXText(JSXText),
    JSXExprContainer(JSXExprContainer),
    JSXElement(Box<JSXElement>),
    // JSXSpreadChild(JSXSpreadChild),
    // JSXFragment(JSXFragment),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSXAttr {
    pub span: Span,
    pub ident: Ident,
    pub value: JSXAttrValue,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum JSXAttrValue {
    Lit(Lit),
    JSXExprContainer(JSXExprContainer),
}
