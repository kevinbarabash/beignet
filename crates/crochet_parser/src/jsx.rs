use chumsky::prelude::*;
use crochet_ast::*;

use crate::util::just_with_padding;

pub fn jsx_parser(
    expr: BoxedParser<'static, char, Expr, Simple<char>>,
) -> BoxedParser<'static, char, Expr, Simple<char>> {
    let str_lit = just("\"")
        .ignore_then(filter(|c| *c != '"').repeated().at_least(1))
        .then_ignore(just("\""))
        .collect::<String>()
        .map_with_span(Lit::str);

    let jsx_text = filter(|c| *c != '<' && *c != '{')
        .repeated()
        .at_least(1)
        .collect::<String>()
        .map_with_span(|value, span| JSXText { span, value });

    let jsx_expr = just_with_padding("{")
        .ignore_then(expr.clone())
        .then_ignore(just_with_padding("}"))
        .map_with_span(|expr, span| JSXExprContainer { span, expr });

    let jsx_attr = text::ident()
        .map_with_span(|name, span| Ident { name, span })
        .then_ignore(just_with_padding("="))
        .then(choice((
            str_lit.map(JSXAttrValue::Lit),
            jsx_expr.clone().map(JSXAttrValue::JSXExprContainer),
        )))
        .map_with_span(|(name, value), span| JSXAttr {
            span,
            ident: name,
            value,
        })
        .padded();

    let jsx_head = just_with_padding("<")
        .ignore_then(text::ident().padded())
        .then(jsx_attr.clone().repeated())
        .then_ignore(just_with_padding(">"));

    let jsx_tail = just("<")
        .ignore_then(just("/"))
        .ignore_then(text::ident().padded())
        .then_ignore(just_with_padding(">"))
        .labelled("jsx_tail");

    let jsx_element_self_closing = just_with_padding("<")
        .ignore_then(text::ident().padded())
        .then(jsx_attr.clone().repeated())
        .then_ignore(just_with_padding("/>"))
        .map_with_span(|(head, attrs), span| JSXElement {
            span,
            name: head,
            attrs,
            children: vec![],
        });

    let jsx_element_child = recursive(|jsx_element_child| {
        // TODO: dedupe with `jsx_element` below
        let jsx_element = jsx_head
            .clone()
            .then(jsx_element_child.repeated())
            .then(jsx_tail.clone())
            .map_with_span(|(((head, attrs), children), tail), span| {
                if head != tail {
                    panic!("JSX head and tail elements must match")
                }

                JSXElementChild::JSXElement(Box::from(JSXElement {
                    span,
                    name: head,
                    attrs,
                    children,
                }))
            });

        choice((
            jsx_element,
            jsx_element_self_closing
                .clone()
                .map(|elem| JSXElementChild::JSXElement(Box::from(elem))),
            jsx_expr.clone().map(JSXElementChild::JSXExprContainer),
            jsx_text.map(JSXElementChild::JSXText),
        ))
    });

    // TODO: dedupe with `jsx_element` above
    let jsx_element = jsx_head
        .then(jsx_element_child.repeated())
        .then(jsx_tail)
        .map_with_span(|(((head, attrs), children), tail), span| {
            if head != tail {
                panic!("JSX head and tail elements must match")
            }

            Expr::JSXElement(JSXElement {
                span,
                name: head,
                attrs,
                children,
            })
        });

    let jsx_parser = choice((jsx_element, jsx_element_self_closing.map(Expr::JSXElement)))
        .labelled("jsx_parser");

    jsx_parser.boxed()
}
