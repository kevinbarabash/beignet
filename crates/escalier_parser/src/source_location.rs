use std::cmp::{max, min};
use std::ops::Range;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Position {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

pub type Span = Range<usize>;

pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

pub fn merge_spans(left: &Span, right: &Span) -> Span {
    Span {
        start: min(left.start, right.start),
        end: max(left.end, right.end),
    }
}
