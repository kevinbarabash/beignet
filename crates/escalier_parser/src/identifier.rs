use crate::source_location::SourceLocation;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Ident {
    pub name: String,
    pub loc: SourceLocation,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BindingIdent {
    pub name: String,
    pub loc: SourceLocation,
    pub mutable: bool,
}
