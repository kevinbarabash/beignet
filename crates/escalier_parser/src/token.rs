use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
