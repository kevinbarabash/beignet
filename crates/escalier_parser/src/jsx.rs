use crate::expr::Expr;
use crate::identifier::Ident;
use crate::span::Span;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum JSXElementName {
    Ident(Ident),
    JSXMemberExpr(JSXMemberExpr),
    // JSXNamespacedName(JSXNamespacedName),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXMemberExpr {
    pub obj: JSXObject,
    pub prop: Ident,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum JSXObject {
    JSXMemberExpr(Box<JSXMemberExpr>),
    Ident(Ident),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXAttr {
    pub name: String,
    pub value: Option<JSXAttrValue>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum JSXAttrValue {
    Str(String),
    ExprContainer(JSXExprContainer),
    // JSX supports using JSXElements and JSXFragments as attribute values without
    // wrapping them in braces.  The spec supports this, but I feel like it's
    // harder to read.
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXExprContainer {
    // pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXOpeningElement {
    pub name: JSXElementName,
    pub attrs: Vec<JSXAttr>, // TODO: support spread
    pub self_closing: bool,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXClosingElement {
    pub name: JSXElementName,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXElement {
    pub span: Span,
    pub opening: JSXOpeningElement,
    pub children: Vec<JSXElementChild>,
    pub closing: Option<JSXClosingElement>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXFragment {
    pub span: Span,
    pub opening: JSXOpeningFragment,
    pub children: Vec<JSXElementChild>,
    pub closing: JSXClosingFragment,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXOpeningFragment {
    // pub span: Span,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXClosingFragment {
    // pub span: Span,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum JSXElementChild {
    Text(JSXText),
    ExprContainer(JSXExprContainer),
    SpreadChild(JSXSpreadChild),
    Element(Box<JSXElement>),
    Fragment(Box<JSXFragment>),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXSpreadChild {
    // pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct JSXText {
    pub span: Span,
    pub value: String,
    // pub raw: String,
}
