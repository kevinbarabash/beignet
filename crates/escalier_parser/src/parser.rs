use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
