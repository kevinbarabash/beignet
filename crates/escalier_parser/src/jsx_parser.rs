use crate::identifier::Ident;
use crate::jsx::*;
use crate::parser::*;
use crate::source_location::*;
use crate::token::{Token, TokenKind};

impl<'a> Parser<'a> {
    pub fn parse_jsx_element(&mut self) -> JSXElement {
        assert_eq!(self.scanner.pop(), Some('<'));
        let name_token = self.lex_ident_or_keyword();
        let name = match name_token.kind {
            TokenKind::Identifier(name) => JSXElementName::Ident(Ident {
                name,
                loc: SourceLocation {
                    start: name_token.loc.start,
                    end: name_token.loc.end,
                },
            }),
            _ => panic!("Expected identifier or keyword"),
        };

        let mut attrs = vec![];
        let mut self_closing = false;

        while !self.scanner.is_done() {
            match self.scanner.peek(0).unwrap() {
                '/' => {
                    self.scanner.pop();
                    assert_eq!(self.scanner.pop(), Some('>'));
                    self_closing = true;
                    break;
                }
                '>' => {
                    self.scanner.pop();
                    break;
                }
                ' ' => {
                    self.scanner.pop();
                }
                _ => {
                    attrs.push(self.parse_jsx_attribute());
                }
            }
        }

        let opening = JSXOpeningElement {
            name,
            attrs,
            self_closing,
        };

        let mut children = vec![];

        let closing = if self_closing {
            None
        } else {
            children = self.parse_jsx_children();

            assert_eq!(self.scanner.pop(), Some('<'));
            assert_eq!(self.scanner.pop(), Some('/'));
            let end_name = self.lex_ident_or_keyword();
            assert_eq!(self.scanner.pop(), Some('>'));

            Some(JSXClosingElement {
                name: match end_name.kind {
                    TokenKind::Identifier(name) => JSXElementName::Ident(Ident {
                        name,
                        loc: SourceLocation {
                            start: end_name.loc.start,
                            end: end_name.loc.end,
                        },
                    }),
                    _ => panic!("Expected identifier or keyword"),
                },
            })
        };

        JSXElement {
            opening,
            children,
            closing,
        }
    }

    pub fn parse_jsx_fragment(&mut self) -> JSXFragment {
        assert_eq!(self.scanner.pop(), Some('<'));
        assert_eq!(self.scanner.pop(), Some('>'));
        let children = self.parse_jsx_children();
        assert_eq!(self.scanner.pop(), Some('<'));
        assert_eq!(self.scanner.pop(), Some('/'));
        assert_eq!(self.scanner.pop(), Some('>'));

        todo!();
        // JSXFragment { children }
    }

    pub fn parse_jsx_attribute(&mut self) -> JSXAttr {
        let name = self.lex_ident_or_keyword();

        let name = match name.kind {
            TokenKind::Identifier(name) => name,
            _ => panic!("Unexpected token"),
        };

        if let Some('=') = self.scanner.peek(0) {
            self.scanner.pop();
        } else {
            return JSXAttr { name, value: None };
        }

        match self.scanner.peek(0).unwrap() {
            '"' => {
                let value = self.lex_string();
                let value = match value.kind {
                    TokenKind::StrLit(value) => value,
                    _ => panic!("Unexpected token"),
                };

                JSXAttr {
                    name,
                    value: Some(JSXAttrValue::Str(value)),
                }
            }
            '{' => {
                self.scanner.pop();

                self.brace_counts.push(0);
                let expr = self.parse_expr();
                self.brace_counts.pop();

                JSXAttr {
                    name,
                    value: Some(JSXAttrValue::ExprContainer(JSXExprContainer {
                        expr: Box::new(expr),
                    })),
                }
            }
            _ => panic!("Unexpected character"),
        }
    }

    pub fn parse_jsx_children(&mut self) -> Vec<JSXElementChild> {
        let mut children = vec![];

        while !self.scanner.is_done() {
            match self.scanner.peek(0).unwrap() {
                '<' => {
                    // TODO: skip of whitespace
                    if self.scanner.peek(1) == Some('/') {
                        break;
                    } else {
                        let elem = self.parse_jsx_element();
                        children.push(JSXElementChild::Element(Box::new(elem)));
                    }
                }
                '{' => {
                    self.scanner.pop();

                    self.brace_counts.push(0);
                    let expr = self.parse_expr();
                    self.brace_counts.pop();

                    children.push(JSXElementChild::ExprContainer(JSXExprContainer {
                        expr: Box::new(expr),
                    }));
                }
                _ => {
                    let text = self.parse_jsx_text();
                    children.push(JSXElementChild::Text(text));
                }
            }
        }

        children
    }

    pub fn parse_jsx_text(&mut self) -> JSXText {
        let start = self.scanner.position();

        let mut value = String::new();

        while !self.scanner.is_done() {
            match self.scanner.peek(0).unwrap() {
                '{' => {
                    break;
                }
                '<' => {
                    break;
                }
                _ => {
                    value.push(self.scanner.pop().unwrap());
                }
            }
        }

        JSXText {
            value,
            // loc: SourceLocation {
            //     start,
            //     end: self.scanner.position(),
            // },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_jsx_element() {
        let mut parser = Parser::new(r#"<Foo bar="baz" qux></Foo>"#);

        let jsx_elem = parser.parse_jsx_element();

        insta::assert_debug_snapshot!(jsx_elem);
    }

    #[test]
    fn parse_self_closing_jsx_element() {
        let mut parser = Parser::new(r#"<Foo bar="baz" qux />"#);

        let jsx_elem = parser.parse_jsx_element();

        insta::assert_debug_snapshot!(jsx_elem);
    }

    #[test]
    fn parse_jsx_element_with_children_text() {
        let mut parser = Parser::new(r#"<h1>Hello, world!</h1>"#);

        let jsx_elem = parser.parse_jsx_element();

        insta::assert_debug_snapshot!(jsx_elem);
    }

    #[test]
    fn parse_jsx_element_with_text_and_exprs() {
        let mut parser = Parser::new(r#"<h1>Hello, {name}!</h1>"#);

        let jsx_elem = parser.parse_jsx_element();

        insta::assert_debug_snapshot!(jsx_elem);
    }

    #[test]
    fn parse_jsx_element_with_children_elements() {
        let mut parser = Parser::new(r#"<ul><li>one</li><li>two</li></ul>"#);

        let jsx_elem = parser.parse_jsx_element();

        insta::assert_debug_snapshot!(jsx_elem);
    }
}
