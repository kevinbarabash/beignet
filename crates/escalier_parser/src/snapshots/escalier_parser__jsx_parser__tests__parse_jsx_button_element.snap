---
source: crates/escalier_parser/src/jsx_parser.rs
expression: jsx_elem
---
JSXElement {
    span: 0..37,
    opening: JSXOpeningElement {
        name: Ident(
            Ident {
                name: "Button",
                span: 1..7,
            },
        ),
        attrs: [
            JSXAttr {
                name: "count",
                value: Some(
                    ExprContainer(
                        JSXExprContainer {
                            expr: Expr {
                                kind: Num(
                                    Num {
                                        value: "5",
                                    },
                                ),
                                span: 15..16,
                                inferred_type: None,
                            },
                        },
                    ),
                ),
            },
            JSXAttr {
                name: "foo",
                value: Some(
                    Str(
                        "bar",
                    ),
                ),
            },
        ],
        self_closing: false,
    },
    children: [],
    closing: Some(
        JSXClosingElement {
            name: Ident(
                Ident {
                    name: "Button",
                    span: 28..37,
                },
            ),
        },
    ),
}
